upgrading/topics/keycloak/changes-19_0_0.adoc

```diff
@@ -0,0 +1,182 @@
+// session.getProvider(UserCache.class) might return null, null-check omitted for brevity.
+session.**getProvider(UserCache.class)**.evict(realm, user);
+----
+
+To trigger the invalidation of a realm, instead of using the `UserCache` API, consider triggering an event:
+
+.Before migration: code will not compile due to a changed return type
+[source,java,subs="+quotes"]
+----
+UserCache cache = session.getProvider(UserCache.class);
+if (cache != null) cache.clear();
+----
+
+.After migration: use the API directly
+[source,java,subs="+quotes"]
+----
+session.invalidate(InvalidationHandler.ObjectType.REALM, realm.getId());
+----
+
+== Credential management for users
+
+Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)` are now managed with `user.credentialManager()._method_(\...)` which gets this functionality closer to API of users, and won't require a distinction of where the user is localed in regard to realm and storage.
+
+The old APIs have been deprecated, and will only work when the legacy storage is enabled in the deployment.
+The new APIs will work with both old and new storages.
+
+.Before migration: accessing a deprecated API
+[source,java,subs="+quotes"]
+----
+session.userCredentialManager()**.createCredential**(realm, user, credentialModel)
+----
+
+.After migration: accessing the new API
+[source,java,subs="+quotes"]
+----
+user.credentialManager()**.createStoredCredential**(credentialModel)
+----
+
+For a custom `UserStorageProvider`, there is a new method `credentialManager()` that needs to be implemented when returning a `UserModel`.
+As those providers run in an environment with the legacy storage enabled, those must return an instance of the `LegacyUserCredentialManager`:
+
+.Before migration: code will not compile due to the new method `credentialManager()` required by `UserModel`
+[source,java,subs="+quotes"]
+----
+public class MyUserStorageProvider implements UserLookupProvider, ... {
+    /* ... */
+    protected UserModel createAdapter(RealmModel realm, String username) {
+        return new AbstractUserAdapter(session, realm, model) {
+            @Override
+            public String getUsername() {
+                return username;
+            }
+        };
+    }
+}
+----
+
+.After migration: implementation of the API `UserModel.credentialManager()` for the legacy store.
+[source,java,subs="+quotes"]
+----
+public class MyUserStorageProvider implements UserLookupProvider, ... {
+    /* ... */
+    protected UserModel createAdapter(RealmModel realm, String username) {
+        return new AbstractUserAdapter(session, realm, model) {
+            @Override
+            public String getUsername() {
+                return username;
+            }
+
+            @Override
+            public SubjectCredentialManager credentialManager() {
+                return new LegacyUserCredentialManager(session, realm, this);
+            }
+        };
+    }
+}
+----
+
```

Comment:
One more item should be added there - Changes in `RealmModel`, specifically that `getUserStorageProviders` and `getUserStorageProvidersStream` methods are removed, and `UserStorageUtil.getUserStorageProvidersStream` needs to be used instead.

Reply:
done: 2314f72399fa8ffcc9b7c5ab2b5e051fc6c7eeee